storybookjs/storybook
# NextJS: Fix Image Context reuse (ensure singleton by externalizing it)

## What I did
I'm not exactly sure when but between https://github.com/storybookjs/storybook/pull/21909 being merged and it being released props specified in `parameters.nextjs.image` have stopped being applied to `next/image`.

After some debugging it turns out that the React Context (`ImageContext`) being used to pass the props from `parameters.nextjs.image` to `next/image` isn't being reused, instead it's recreated in every file where it's imported. This happens because webpack inlines the contents of `context.ts` instead of keeping the import statement intact.

Since we want to reuse the same context everywhere we should make sure webpack preserves the import even after building, which can be done by marking `context.ts` as an external file that shouldn't be included in the output bundle.

Note we also need to mark `decorator.tsx` as external, or else it will be included in the `preview.js` bundle, causing the import of `context.ts` to fail.

## How to test

If you look at the current output you can see `ImageContext` being recreated in every file it's imported.
For example in `dist/preview.js`, `dist/images/next-image.js` or  `dist/images/next-legacy-image.js` you will see the following line:
```js
ImageContext=(0,import_react.createContext)({})
```

After building this branch the output instead turns into
```js
var import_decorator = require("./images/decorator");
```
for `preview.js` and 
```js
var import_context = require("./context");
```
for `decorator.js` and finally
```js
var import_context = require("./context");
```
for `next-image.js`, `next-legacy-image.js`, and` next-future-image.js`, which means the context is correctly being reused instead of recreated.

## Checklist

      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers


`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`
